---
title: "DE_permutations.Rmd"
author: "Kate Lee"
date: "05/08/2019"
output: html_document
---

AIM: 
- take total number of expressed genes and total number of significantly differentially expressed genes for each species
- randomise the DE genes multiple times (e.g. 1 million)
- check to see if there are overlaps of DE genes between species 

Adapted from https://mac-theobio.github.io/QMEE/permutation_examples.html
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

Read in the data from the core gene set analysis (this shows a line for every ortholog described in the gene set)

```{r get_data}

inf_ps_rfmt <- read.delim("../core_gene_sets/core_genes_INF_PS_rfmt.txt", header = TRUE, sep = "\t")
head(inf_ps_rfmt)
# remove lines with no fold change reading
inf_ps <- inf_ps_rfmt[!is.na(inf_ps_rfmt$log2fc),]

# remove any duplicate rows
inf_ps <- inf_ps %>% distinct

# set DE at 1 for log2FC >= 1 and svalue <= 0.005, set DE at 2 for log2FC >= 2 and svalue <= 0.005
inf_ps <- inf_ps %>% mutate(DE = ifelse((log2fc >= 2) &(svalue_1 <= 0.005), 2, 
                                        ifelse((log2fc >= 1) &(svalue_1 <= 0.005), 1, 0)),
                            DE2 = ifelse((log2fc >= 1) &(svalue_1 <= 0.005), 1, 0), 
                            DE4 = ifelse((log2fc >= 2) &(svalue_1 <= 0.005), 1, 0))

# flag direction of fold change
inf_ps <- inf_ps %>% mutate(dir_FC = ifelse((log2fc < 0), "-", "+"))

# subset data
inf_ps <- inf_ps[, c("species", "orthogroup", "gene_id", "dir_FC", "DE", "DE2", "DE4")]

# total genes and total number of DE genes per species
inf_ps %>% group_by(species) %>% summarise(total = n(), total_DE2 = sum(DE2), total_DE4 = sum(DE4))

# total number of ortholog overlaps in observed data
inf_ps <- inf_ps %>% 
          group_by(orthogroup) %>% 
          mutate(number_spp = length(unique(species)), 
                 FC2_overlaps = ifelse((length(unique(species)) > 1) & (sum(DE2) == length(unique(species))), 1, 0),        
                 FC4_overlaps = ifelse((length(unique(species)) > 1) & (sum(DE4) == length(unique(species))), 1, 0), 
                 core_overlaps = ifelse((length(unique(species)) > 1) & (sum(DE2) == (length(unique(species)))) & (sum(DE4) >= 1), 1, 0))
```



```{r permute_data}
set.seed(101) ## for reproducibility
nsim <- 1000000
FC4_up <- numeric(nsim) ## set aside space for results
FC4_down <- numeric(nsim)
FC2_up <- numeric(nsim)
FC2_down <- numeric(nsim)
core_up <- numeric(nsim)
core_down <- numeric(nsim)
results <- data.frame()
for (i in 1:nsim) {
    ## standard approach: scramble response value
    bdat <- data.frame()
    for (spp in unique(inf_ps_rfmt$species)) {
      perm <- sample(nrow(inf_ps[inf_ps$species == spp,]))
      tmp <- transform(inf_ps[inf_ps$species == spp,], DE = DE[perm])
      tmp <- transform(tmp, DE2 = ifelse((DE == 1) | (DE == 2), 1, 0), DE4 = ifelse((DE == 2), 1, 0))
      bdat <- rbind(bdat, tmp)
    }
    ## check the number of ortholog overlaps and store them
    res <- bdat %>% group_by(orthogroup) %>% 
           mutate(number_spp = length(unique(species)), 
                 FC2_overlaps = ifelse((length(unique(species)) > 1) & (sum(DE2) == length(unique(species))), 1, 0),        
                 FC4_overlaps = ifelse((length(unique(species)) > 1) & (sum(DE4) == length(unique(species))), 1, 0), 
                 core_overlaps = ifelse((length(unique(species)) > 1) & (sum(DE2) == (length(unique(species)))) & (sum(DE4) >= 1), 1, 0))
    FC4_up[i] <- nrow(filter(res, (FC4_overlaps == 1) & (dir_FC == "+")))
    FC4_down[i] <- nrow(filter(res, (FC4_overlaps == 1) & (dir_FC == "-")))
    FC2_up[i] <- nrow(filter(res, (FC2_overlaps == 1) & (dir_FC == "+")))
    FC2_down[i] <- nrow(filter(res, (FC2_overlaps == 1) & (dir_FC == "-")))
    core_up[i] <- nrow(filter(res, (core_overlaps == 1) & (dir_FC == "+")))
    core_down[i] <- nrow(filter(res, (core_overlaps == 1) & (dir_FC == "-")))
    tmp <- c(FC4_up[i], FC4_down[i], FC2_up[i], FC2_down[i], core_up[i], core_down[i])
    results <- rbind(results, tmp)
}
colnames(results) <- c("FC2_up", "FC2_down", "FC4_up", "FC4_down", "core_up", "core_down")

write.table(results, "randomised_DE_results_.txt")
```

```{r check_results}
FC4_up
FC4_down
FC2_up
FC2_down
core_up
core_down
# observed results
FC2_obs_up <- nrow(filter(inf_ps, (FC2_overlaps == 1) & (dir_FC == "+")))
FC2_obs_down <- nrow(filter(inf_ps, (FC2_overlaps == 1) & (dir_FC == "-")))
FC4_obs_up <- nrow(filter(inf_ps, (FC4_overlaps == 1) & (dir_FC == "+")))
FC4_obs_down <- nrow(filter(inf_ps, (FC4_overlaps == 1) & (dir_FC == "-")))
core_obs_up <- nrow(filter(inf_ps, (core_overlaps == 1) & (dir_FC == "+")))
core_obs_down <- nrow(filter(inf_ps, (core_overlaps == 1) & (dir_FC == "-")))

pvalue_FC2_up <- ((nrow(results[results$FC2_up >= FC2_obs_up, ])+1)/(nsim +1))
pvalue_FC2_down <- ((nrow(results[results$FC2_down <= FC2_obs_down, ])+1)/(nsim +1))
pvalue_FC4_up <- ((nrow(results[results$FC4_up >= FC4_obs_up, ])+1)/(nsim +1))
pvalue_FC4_down <- ((nrow(results[results$FC4_down <= FC4_obs_down, ])+1)/(nsim +1))
pvalue_core_up <- ((nrow(results[results$core_up >= core_obs_up, ])+1)/(nsim +1))
pvalue_core_down <- ((nrow(results[results$core_down <= core_obs_down, ])+1)/(nsim +1))

pvalue_FC2_up
pvalue_FC2_down
pvalue_FC4_up
pvalue_FC4_down
pvalue_core_up
pvalue_core_down

```


```{r graph_permutations}


ggplot(results) +
  geom_bar(aes(x = FC2_up)) +
  geom_vline(xintercept = FC2_obs_up, color = "red", size=1) 

ggplot(results) +
  geom_bar(aes(x = FC2_down)) +
  geom_vline(xintercept = FC2_obs_down, color = "red", size=1)


ggplot(results) +
  geom_bar(aes(x = FC4_up)) +
  geom_vline(xintercept = FC4_obs_up, color = "red", size=1) 

ggplot(results) +
  geom_bar(aes(x = FC4_down)) +
  geom_vline(xintercept = FC4_obs_down, color = "red", size=1)

ggplot(results) +
  geom_bar(aes(x = core_up)) +
  geom_vline(xintercept = core_obs_up, color = "red", size=1) 

ggplot(results) +
  geom_bar(aes(x = core_down)) +
  geom_vline(xintercept = core_obs_down, color = "red", size=1)


```