---
title: "E.elymi_DEseq2_PS_STR"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)


# DESeq2 installation
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("DESeq2", version = "3.8")

# tximport installation
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("tximport", version = "3.8")

# tximportData installation to see example data
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("tximportData", version = "3.8")

#install.packages("readr")

#install.packages("tidyverse")


library("DESeq2")
library("tximport")
library("tximportData")
library("readr")
library("tidyverse")
```

Following example from https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html
http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

```{r get_data}

# POP = speices used
# Center = sequencing center
# assay = sequencing
# run_number = 

# import sample list and assign column and row names
samples <- read.table("../../config.txt", header = FALSE)
colnames(samples) <- c("pop", "center", "assay", "run_number", "lane", "sample", "experiment", "run", "condition")
rownames(samples) <- samples$run

# subset only one species
samples <- samples[samples$pop == "E.typhina_E8",]
samples
```


```{r get_quant_data}

# local files
#dir <- "../../2_salmon"

# files on external hard drive
dir <- "/mnt/usb-Seagate_Expansion_NAA7H6EA-0:0-part2/Epichloe/2_salmon"


# get salmon quant.sf files for each fq file
files <- file.path(dir, "salmon_quant", samples$pop, samples$run, "quant.sf")
names(files) <- samples$run
files
```


```{r get_transcript_and_gene_names}

tx2gene <- read.table(file.path(dir, "transcriptome/E.typhina_E8/transcriptome.headers.txt"))
colnames(tx2gene) <- c("TXNAME", "GENEID")
#head(tx2gene)
```


```{r}
# read in count data
txi <- tximport(files, type="salmon", tx2gene=tx2gene)
```
 

```{r check_txi_output, eval=FALSE, include=FALSE}
write.table(txi, "temp_txi_output_check.txt", quote = FALSE, sep="\t")
```


```{r}
# import data to DESeq
ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = samples,
                                   design = ~ condition)

```

DESeq2 Analysis


filter out data with less than 10 read counts - skip this, Barry and Daniel want to see reads that are present in one sample and not in another (most of the secondary metabolites show this pattern)
```{r filter_out_low_counts}

#keep <- rowSums(counts(ddsTxi)) >= 10
#ddsTxi <- ddsTxi[keep,]
```

regularized-logarithm transformation or rlog (Love, Huber, and Anders 2014)
specified blind = FALSE, which means that differences between cell lines and treatment (the variables in the design) will not contribute to the expected variance-mean trend of the experiment.
```{r rlog_trans}

rld <- rlog(ddsTxi, blind = FALSE)
head(assay(rld), 3)
```

```{r vst_log_trans}

vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)
```


```{r see_rlog_trans_effects}
library("dplyr")
library("ggplot2")
#install.packages("hexbin")
library("hexbin")

dds <- estimateSizeFactors(dds)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  #as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))
  
colnames(df)[1:2] <- c("x", "y")  

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)  

```

```{r sample_distances}

sampleDists <- dist(t(assay(rld)))
sampleDists
```

```{r heatmap_of_sample_dist}
#install.packages("pheatmap")
#install.packages("RColorBrewer")
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
#rownames(sampleDistMatrix) <- paste( vsd$dex, vsd$cell, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)



```



if you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels. Here just set the reference condition
```{r set_ref_condition}

ddsTxi$condition <- relevel(ddsTxi$condition, ref = "PS")
```


```{r run_DESeq}

dds <- DESeq(ddsTxi)
res <- results(dds)
res
```

```{r diff_gene_exp}
summary(res) #summary of results
```


Note padj => DESeq2 uses the Benjamini-Hochberg (BH) adjustment (Benjamini and Hochberg 1995) as implemented in the base R p.adjust function.
if we consider a fraction of 10% false positives acceptable, we can consider all genes with an adjusted p value below 10% = 0.1 as significant.
```{r summary_sorted_by_p, eval=FALSE, include=FALSE}

res <- res[order(res$padj),]
head(res)
```


```{r plot_top_siz_diffs, eval=FALSE, include=FALSE}

top_counts <- rownames(res)
par(mfrow=c(2,3))
plotCounts(dds, gene=top_counts[1], intgroup="condition")
plotCounts(dds, gene=top_counts[2], intgroup="condition")
plotCounts(dds, gene=top_counts[3], intgroup="condition")
plotCounts(dds, gene=top_counts[4], intgroup="condition")
plotCounts(dds, gene=top_counts[5], intgroup="condition")
plotCounts(dds, gene=top_counts[6], intgroup="condition")
```


```{r }

#reset par
par(mfrow=c(1,1))
# Make a basic volcano plot
with(res, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-3,3)))

# Add colored points: blue if padj<0.01, red if log2FC>1 and padj<0.05)
with(subset(res, padj<.01 ), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))
with(subset(res, padj<.01 & abs(log2FoldChange)>2), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
```

```{r PCA}
#First we need to transform the raw count data
#vst function will perform variance stabilizing transformation

vsdata <- vst(dds, blind=FALSE)

plotPCA(vsdata, intgroup="condition") #using the DESEQ2 plotPCA fxn we can
```


Get data from gff3 file
ID genes that are over 2 fold change that are close to one another

```{r ID_clusters}

positions <- read.table("../../0_raw_data/genome_assembly/Ety_8/Epichloe_typhina_E8.gff3", header = FALSE, comment.char = "#", blank.lines.skip = TRUE, fill = TRUE, sep = "\t")
colnames(positions) <- c("contig", "source", "type", "start", "end", "score", "strand", "phase", "attributes")

# select out mRNA (these are the ones with product described)
mRNA <- positions[positions$type == "mRNA",]

# split attributes to get c("trans_id", "gene_id", "product")
temp <- str_split(mRNA$attributes, ";", n=3)
df <- as.data.frame(t(as.data.frame(temp)))
rownames(df)<-NULL
mRNA[, c("trans_id", "gene_id", "product")] <- df
mRNA$trans_id <- str_replace(mRNA$trans_id, "ID=", "")
mRNA$gene_id <- str_replace(mRNA$gene_id, "Parent=", "")
mRNA$product <- str_replace(mRNA$product, "product=", "")


# merge DESeq output (res) and GFF3 data for mRNA (where the annotations are)
res$gene_id <- rownames(res)
tempdf <- as.data.frame(res)
GOI <- merge(tempdf, mRNA, by="gene_id")
GOI$score <- NULL
GOI$strand <- NULL
GOI$phase <- NULL
GOI$attributes <- NULL
# subset target results that have an adjusted pvalue less than .01 and a log2FoldChange >2
targets <- rownames(subset(res, padj<.01 & abs(log2FoldChange)>1))

GOI <- GOI[GOI$gene_id %in% targets,]
#table <- GOI %>% group_by(seqid, start)
#table$attributes <- NULL

str(GOI)
head(GOI)
write.table(GOI, "E.typhina_PS_STR_log2foldchange.txt", quote = FALSE, row.names = FALSE, sep = "\t")

```