---
title: "core_gene_set"
author: "Kate"
date: "10 April 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("tidyverse")
library("tidyverse")
```


Take in results from setB mutant vs WT and clrD mutant vs WT
```{r get_data}
elymi <- read.delim("E.elymi_NfE728/result_tables/E.elymi_NfE728_STR_PS.txt", header = TRUE)
festucae <- read.delim("E.festucae_E2368/result_tables/E.festucae_E2368_STR_PS.txt", header = TRUE)
typhina <- read.delim("E.typhina_E8/result_tables/E.typhina_E8_STR_PS.txt", header = TRUE)
```


```{r subset_apeglm_results}
keep <- c("gene_id","PS_fpkm", "STR_fpkm", "apeglm_1_log2FC", "apeglm_1_lfcSE",  "apeglm_1_svalue", "apeglm_2_log2FC", "apeglm_2_lfcSE", "apeglm_2_svalue")
elymi <- elymi[, keep]
festucae <- festucae[, keep]
typhina <- typhina[, keep]

# set colnames for each set of results
colnames(elymi) <- paste("elymi", colnames(elymi), sep = "_")
colnames(festucae) <- paste("festucae", colnames(festucae), sep = "_")
colnames(typhina) <- paste("typhina", colnames(typhina), sep = "_")
```


```{r rename_genes_by_ortholog_names}

orthology <- read.table("../orthology/orthology_groups.tsv", header = TRUE)

orthology$Eel <- gsub("-T1", "", orthology$Eel)
orthology$Efe <- gsub("-T1", "", orthology$Efe)
orthology$Ety <- gsub("-T1", "", orthology$Ety)

elymi <- merge(elymi, orthology[, c("group", "Eel")], by.x = "elymi_gene_id", by.y = "Eel")
festucae <- merge(festucae, orthology[, c("group", "Efe")], by.x = "festucae_gene_id", by.y = "Efe")
typhina <- merge(typhina, orthology[, c("group", "Ety")], by.x = "typhina_gene_id", by.y = "Ety")

#head(elymi)
#str(orthology)    #7805 obs of 9 variables
#str(elymi)
#str(festucae)
#str(typhina)
```


```{r merge_data}

# merge result tables
core_set <- merge(elymi, festucae, by = "group", all = TRUE)
core_set <- merge(core_set, typhina, by = "group", all = TRUE)

rownames(core_set) <- core_set$group
colnames(core_set)
str(core_set)
```


```{r flag_apeglm_normalised_log2FC}

# flag if up by 4fold in one data set and by up by at least 2 fold in the other two
targets_up <-  rownames(subset(core_set, ( ((elymi_apeglm_2_svalue <.005 & elymi_apeglm_2_log2FC >= 2 ) 
                                          | (festucae_apeglm_2_svalue <.005 & festucae_apeglm_2_log2FC >= 2 ) 
                                          | (typhina_apeglm_2_svalue <.005 & typhina_apeglm_2_log2FC >= 2 ) )
                                          & (elymi_apeglm_1_svalue <.005 & elymi_apeglm_1_log2FC >= 1 ) 
                                          & (festucae_apeglm_1_svalue <.005 & festucae_apeglm_1_log2FC >= 1 ) 
                                          & (typhina_apeglm_1_svalue <.005 & typhina_apeglm_1_log2FC >= 1 ) )))
                                            
core_set$core_up <- ifelse(rownames(core_set) %in% targets_up, 1, 0)

targets_down <-  rownames(subset(core_set, ( ((elymi_apeglm_2_svalue <.005 & elymi_apeglm_2_log2FC <= -2 ) 
                                          | (festucae_apeglm_2_svalue <.005 & festucae_apeglm_2_log2FC <= -2 ) 
                                          | (typhina_apeglm_2_svalue <.005 & typhina_apeglm_2_log2FC <= -2 ) )
                                          & (elymi_apeglm_1_svalue <.005 & elymi_apeglm_1_log2FC <= -1 ) 
                                          & (festucae_apeglm_1_svalue <.005 & festucae_apeglm_1_log2FC <= -1 ) 
                                          & (typhina_apeglm_1_svalue <.005 & typhina_apeglm_1_log2FC <= -1 ) )))

core_set$core_down <- ifelse(rownames(core_set) %in% targets_down, 1, 0)



##### 4 fold change
targets_topup <-  rownames(subset(core_set, ( (elymi_apeglm_2_svalue <.005 & elymi_apeglm_2_log2FC >= 2 ) 
                                          & (festucae_apeglm_2_svalue <.005 & festucae_apeglm_2_log2FC >= 2 ) 
                                          & (typhina_apeglm_2_svalue <.005 & typhina_apeglm_2_log2FC >= 2 ) )))

core_set$top_4FC_up <- ifelse(rownames(core_set) %in% targets_topup, 1, 0)

targets_topdown <-  rownames(subset(core_set, ( (elymi_apeglm_2_svalue <.005 & elymi_apeglm_2_log2FC <= -2 ) 
                                          & (festucae_apeglm_2_svalue <.005 & festucae_apeglm_2_log2FC <= -2 ) 
                                          & (typhina_apeglm_2_svalue <.005 & typhina_apeglm_2_log2FC <= -2 ) )))

core_set$top_4FC_down <- ifelse(rownames(core_set) %in% targets_topdown, 1, 0)


##### 2 fold change
targets_topup <-  rownames(subset(core_set, ( (elymi_apeglm_1_svalue <.005 & elymi_apeglm_1_log2FC >= 1 ) 
                                          & (festucae_apeglm_1_svalue <.005 & festucae_apeglm_1_log2FC >= 1 ) 
                                          & (typhina_apeglm_1_svalue <.005 & typhina_apeglm_1_log2FC >= 1 ) )))

core_set$top_2FC_up <- ifelse(rownames(core_set) %in% targets_topup, 1, 0)

targets_topdown <-  rownames(subset(core_set, ( (elymi_apeglm_1_svalue <.005 & elymi_apeglm_1_log2FC <= -1 ) 
                                          & (festucae_apeglm_1_svalue <.005 & festucae_apeglm_1_log2FC <= -1 ) 
                                          & (typhina_apeglm_1_svalue <.005 & typhina_apeglm_1_log2FC <= -1 ) )))

core_set$top_2FC_down <- ifelse(rownames(core_set) %in% targets_topdown, 1, 0)


sum(core_set$core_up)
sum(core_set$core_down)
sum(core_set$top_4FC_up)
sum(core_set$top_4FC_down)
sum(core_set$top_2FC_up)
sum(core_set$top_2FC_down)

write.table(core_set, "STR_PS/core_set_STR_PS.txt", quote = FALSE, col.names = TRUE, row.names = FALSE)

```


# Graph overlaps with UpSetR package
```{r UpSet_diagram_apeglm_normalised_results}
#install.packages("UpSetR")
library(UpSetR)

rownames(core_set) <- core_set$gene_id

listInput <- list(elymi_2Fold_up = rownames(core_set[core_set$elymi_apeglm_1_log2FC >= 1 & core_set$elymi_apeglm_1_svalue < 0.005 & !is.na(core_set$elymi_apeglm_1_log2FC), ]),
                  elymi_4Fold_up = rownames(core_set[core_set$elymi_apeglm_2_log2FC >= 2 & core_set$elymi_apeglm_2_svalue < 0.005 & !is.na(core_set$elymi_apeglm_1_log2FC), ]),
                  festucae_2Fold_up = rownames(core_set[core_set$festucae_apeglm_1_log2FC >= 1 & core_set$festucae_apeglm_1_svalue < 0.005 & !is.na(core_set$festucae_apeglm_1_log2FC), ]),
                  festucae_4Fold_up = rownames(core_set[core_set$festucae_apeglm_2_log2FC >= 2 & core_set$festucae_apeglm_2_svalue < 0.005 & !is.na(core_set$festucae_apeglm_1_log2FC), ]),
                  typhina_2Fold_up = rownames(core_set[core_set$typhina_apeglm_1_log2FC >= 1 & core_set$typhina_apeglm_1_svalue < 0.005 & !is.na(core_set$typhina_apeglm_1_log2FC), ]),
                  typhina_4Fold_up = rownames(core_set[core_set$typhina_apeglm_2_log2FC >= 2 & core_set$typhina_apeglm_2_svalue < 0.005 & !is.na(core_set$typhina_apeglm_1_log2FC), ]),
                  
                  elymi_2Fold_down = rownames(core_set[core_set$elymi_apeglm_1_log2FC <= -1 & core_set$elymi_apeglm_1_svalue < 0.005 & !is.na(core_set$elymi_apeglm_1_log2FC), ]),
                  elymi_4Fold_down = rownames(core_set[core_set$elymi_apeglm_2_log2FC <= -2 & core_set$elymi_apeglm_2_svalue < 0.005 & !is.na(core_set$elymi_apeglm_1_log2FC), ]),
                  festucae_2Fold_down = rownames(core_set[core_set$festucae_apeglm_1_log2FC <= -1 & core_set$festucae_apeglm_1_svalue < 0.005 & !is.na(core_set$festucae_apeglm_1_log2FC), ]),
                  festucae_4Fold_down = rownames(core_set[core_set$festucae_apeglm_2_log2FC <= -2 & core_set$festucae_apeglm_2_svalue < 0.005 & !is.na(core_set$festucae_apeglm_1_log2FC), ]),
                  typhina_2Fold_down = rownames(core_set[core_set$typhina_apeglm_1_log2FC <= -1 & core_set$typhina_apeglm_1_svalue < 0.005 & !is.na(core_set$typhina_apeglm_1_log2FC), ]),
                  typhina_4Fold_down = rownames(core_set[core_set$typhina_apeglm_2_log2FC <= -2 & core_set$typhina_apeglm_2_svalue < 0.005 & !is.na(core_set$typhina_apeglm_1_log2FC), ]),
                  
                  core_up = rownames(core_set[core_set$core_up == 1 & !is.na(core_set$elymi_apeglm_1_log2FC) & !is.na(core_set$festucae_apeglm_1_log2FC) & !is.na(core_set$typhina_apeglm_1_log2FC), ]),
                  core_down = rownames(core_set[core_set$core_down == 1 & !is.na(core_set$elymi_apeglm_1_log2FC) & !is.na(core_set$festucae_apeglm_1_log2FC) & !is.na(core_set$typhina_apeglm_1_log2FC), ]),
                  top_4FC_up = rownames(core_set[core_set$top_4FC_up == 1 & !is.na(core_set$elymi_apeglm_1_log2FC) & !is.na(core_set$festucae_apeglm_1_log2FC) & !is.na(core_set$typhina_apeglm_1_log2FC), ]),
                  top_4FC_down = rownames(core_set[core_set$top_4FC_down == 1 & !is.na(core_set$elymi_apeglm_1_log2FC) & !is.na(core_set$festucae_apeglm_1_log2FC) & !is.na(core_set$typhina_apeglm_1_log2FC), ])
                  
                  )

sets <- names(listInput)
colors <- c("blue", "blue", "blue", "blue", "blue", "blue", "blue", "blue", "purple", "purple", "purple", "purple", "purple", "purple", "purple", "purple")
metadata <- as.data.frame(cbind(sets, colors))

set_colors <- c(typhina_2Fold_down = "darkcyan", festucae_2Fold_down = "darkcyan",  elymi_2Fold_down = "darkcyan", typhina_4Fold_down = "darkcyan", festucae_4Fold_down = "darkcyan", elymi_4Fold_down = "darkcyan",  core_down = "darkblue", top_4FC_down = "darkblue", typhina_2Fold_up = "purple", festucae_2Fold_up = "purple", elymi_2Fold_up = "purple", typhina_4Fold_up = "purple",  festucae_4Fold_up = "purple", elymi_4Fold_up = "purple", core_up = "red", top_4FC_up = "red")    #   shade.color = colors

pdf("STR_PS/UpSet_core_set_STR_PS.pdf", onefile = FALSE)
upset(fromList(listInput), sets = c("core_down", "core_up",  "top_4FC_down", "top_4FC_up"), keep.order = TRUE, order.by = "freq")
dev.off()

pdf("STR_PS/UpSet_all_FC_data_STR_PS.pdf", onefile = FALSE)
upset(fromList(listInput), sets = c("typhina_2Fold_down", "festucae_2Fold_down",  "elymi_2Fold_down", "typhina_4Fold_down", "festucae_4Fold_down", "elymi_4Fold_down",  "core_down","top_4FC_down", "typhina_2Fold_up", "festucae_2Fold_up", "elymi_2Fold_up", "typhina_4Fold_up",  "festucae_4Fold_up", "elymi_4Fold_up", "core_up", "top_4FC_up"), keep.order = TRUE, order.by = "freq", nintersects = 53, set.metadata = list(data = metadata, 
                          plots = list(list(type = "matrix_rows", column = "sets", colors = set_colors, alpha = 0.5))))
dev.off()

pdf("STR_PS/UpSet_all_FC_data_up_STR_PS.pdf", onefile = FALSE)
upset(fromList(listInput), sets = c("typhina_2Fold_up", "festucae_2Fold_up", "elymi_2Fold_up", "typhina_4Fold_up",  "festucae_4Fold_up", "elymi_4Fold_up", "core_up", "top_4FC_up"), keep.order = TRUE, order.by = "freq")
dev.off()

pdf("STR_PS/UpSet_all_FC_data_down_STR_PS.pdf", onefile = FALSE)
upset(fromList(listInput), sets = c("typhina_2Fold_down", "festucae_2Fold_down",  "elymi_2Fold_down", "typhina_4Fold_down", "festucae_4Fold_down", "elymi_4Fold_down",  "core_down","top_4FC_down"), keep.order = TRUE, order.by = "freq")
dev.off()
```
```{r add_ortho_info}

core_set <- merge(core_set, orthology, by = "group")
core_set[, c("Eel", "Efe", "Ety")] <- NULL
colnames(core_set)
str(core_set)
```

```{r add_annoatations}

#### GET ANNOTATIONS FROM annotations.txt file from David's funannotate pipeline file
elymi_ann <- read.delim("../0_raw_data/genome_assembly/Eel_728/Epichloe_elymi.annotations.txt", header = TRUE, comment.char = "#", blank.lines.skip = TRUE, fill = TRUE, sep = "\t")
festucae_ann <- read.delim("../0_raw_data/genome_assembly/Efe_2368/Epichloe_festucae.annotations.txt", header = TRUE, comment.char = "#", blank.lines.skip = TRUE, fill = TRUE, sep = "\t")
typhina_ann <- read.delim("../0_raw_data/genome_assembly/Ety_8/Epichloe_typhina_E8.annotations.txt", header = TRUE, comment.char = "#", blank.lines.skip = TRUE, fill = TRUE, sep = "\t")

elymi_ann$protein_length <- nchar(as.character(elymi_ann$Translation))
festucae_ann$protein_length <- nchar(as.character(festucae_ann$Translation))
typhina_ann$protein_length <- nchar(as.character(typhina_ann$Translation))

# tag each spp annotation
colnames(elymi_ann) <- paste("elymi", colnames(elymi_ann), sep = "_")
colnames(festucae_ann) <- paste("festucae", colnames(festucae_ann), sep = "_")
colnames(typhina_ann) <- paste("typhina", colnames(typhina_ann), sep = "_")

#colnames(elymi_ann)
str(core_set)

 # merge DESeq output (res) and GFF3 data for mRNA (where the annotations are)
core_set <- merge(core_set, typhina_ann, by.x = "typhina_gene_id", by.y = "typhina_GeneID", all.x = TRUE)
core_set <- merge(core_set, festucae_ann, by.x = "festucae_gene_id", by.y = "festucae_GeneID", all.x = TRUE)
core_set <- merge(core_set, elymi_ann, by.x = "elymi_gene_id", by.y = "elymi_GeneID", all.x = TRUE)

#head(core_set)
str(core_set)

write.table(core_set, "STR_PS/core_gene_set_STR_PS_ann.txt", quote = FALSE, row.names = FALSE, sep = "\t")
```

##### ARCHIVED

```{r flag_genes_of_interest, eval=FALSE, include=FALSE}

# flag presence in both data sets
targets <- rownames(subset(core_set, (setB_padj<.01 & setB_log2FC >= 1 & clrD_padj<.01 & clrD_log2FC >= 1 & !is.na(setB_log2FC) & !is.na(clrD_log2FC) )))
core_set$core_log2FC_1_up <- ifelse(rownames(core_set) %in% targets, 1, 0)
targets <- rownames(subset(core_set, (setB_padj<.01 & setB_log2FC >= 2 & clrD_padj<.01 & clrD_log2FC >= 2 & !is.na(setB_log2FC) & !is.na(clrD_log2FC) )))
core_set$core_log2FC_2_up <- ifelse(rownames(core_set) %in% targets, 1, 0)

targets <- rownames(subset(core_set, (setB_padj<.01 & setB_log2FC <= -1 & clrD_padj<.01 & clrD_log2FC <= -1 & !is.na(setB_log2FC) & !is.na(clrD_log2FC) )))
core_set$core_log2FC_1_down <- ifelse(rownames(core_set) %in% targets, 1, 0)
targets <- rownames(subset(core_set, (setB_padj<.01 & setB_log2FC <= -2 & clrD_padj<.01 & clrD_log2FC <= -2 & !is.na(setB_log2FC) & !is.na(clrD_log2FC) )))
core_set$core_log2FC_2_down <- ifelse(rownames(core_set) %in% targets, 1, 0)

# visual check
#colnames(core_set)
#core_set[, c("gene_id", "setB_baseMean", "setB_log2FC", "setB_padj", "setB_log2FC_1", "setB_log2FC_2", "clrD_log2FC",  "clrD_padj", "clrD_log2FC_1", "clrD_log2FC_2", "core_log2FC_1_up", "core_log2FC_2_up", "core_log2FC_1_down", "core_log2FC_2_down")]
#core_set[is.na(core_set$setB_log2FC), c("gene_id",  "setB_log2FC", "setB_log2FC_1", "setB_log2FC_2", "clrD_log2FC",  "clrD_log2FC_1", "clrD_log2FC_2", "core_log2FC_1_up", "core_log2FC_2_up", "core_log2FC_1_down", "core_log2FC_2_down")]
```

# Graph overlaps with UpSetR package
```{r UpSet_diagram_raw_DESeq, eval=FALSE, include=FALSE}
#install.packages("UpSetR")
library(UpSetR)

rownames(core_set) <- core_set$gene_id

listInput <- list(clrD_2Fold_up = rownames(core_set[core_set$clrD_log2FC >= 1 & core_set$clrD_padj < 0.01 & !is.na(core_set$clrD_log2FC), ]),
                  clrD_4Fold_up = rownames(core_set[core_set$clrD_log2FC >= 2 & core_set$clrD_padj < 0.01 & !is.na(core_set$clrD_log2FC), ]),
                  setB_2Fold_up = rownames(core_set[core_set$setB_log2FC >= 1 & core_set$setB_padj < 0.01 & !is.na(core_set$setB_log2FC), ]),
                  setB_4Fold_up = rownames(core_set[core_set$setB_log2FC >= 2 & core_set$setB_padj < 0.01 & !is.na(core_set$setB_log2FC), ]),
                  
                  clrD_2Fold_down = rownames(core_set[core_set$clrD_log2FC <= -1 & core_set$clrD_padj < 0.01 & !is.na(core_set$clrD_log2FC), ]),
                  clrD_4Fold_down = rownames(core_set[core_set$clrD_log2FC <= -2 & core_set$clrD_padj < 0.01 & !is.na(core_set$clrD_log2FC), ]),
                  setB_2Fold_down = rownames(core_set[core_set$setB_log2FC <= -1 & core_set$setB_padj < 0.01 & !is.na(core_set$setB_log2FC), ]),
                  setB_4Fold_down = rownames(core_set[core_set$setB_log2FC <= -2 & core_set$setB_padj < 0.01 & !is.na(core_set$setB_log2FC), ]),
                  
                  core_2Fold_up = rownames(core_set[core_set$core_log2FC_1_up == 1 & !is.na(core_set$setB_log2FC) & !is.na(core_set$clrD_log2FC), ]),
                  core_4Fold_up = rownames(core_set[core_set$core_log2FC_2_up == 1 & !is.na(core_set$setB_log2FC) & !is.na(core_set$clrD_log2FC), ]),
                  core_2Fold_down = rownames(core_set[core_set$core_log2FC_1_down == 1 & !is.na(core_set$setB_log2FC) & !is.na(core_set$clrD_log2FC), ]),
                  core_4Fold_down = rownames(core_set[core_set$core_log2FC_2_down == 1 & !is.na(core_set$setB_log2FC) & !is.na(core_set$clrD_log2FC), ])
                  )


upset(fromList(listInput), sets = c("core_2Fold_down", "core_4Fold_down", "core_2Fold_up", "core_4Fold_up"), keep.order = TRUE, order.by = "freq")


jpeg("plots/UpSet_all_FC_data.jpeg")
upset(fromList(listInput), sets = c("setB_2Fold_down", "setB_4Fold_down", "clrD_2Fold_down", "clrD_4Fold_down", "core_2Fold_down", "core_4Fold_down", "setB_2Fold_up", "setB_4Fold_up", "clrD_2Fold_up", "clrD_4Fold_up", "core_2Fold_up", "core_4Fold_up"), keep.order = TRUE, order.by = "freq")
dev.off()

```